### ISSUE TITLE
[pymc3] Gamma(1,1) MAP issue

### ISSUE THREAD

**Issue by kforeman** [pymc-devs/pymc3#334]
I've found that when I run find_MAP() on a Gamma(1,1) distribution it ends up with negative value (and thus -Inf logp).

Here's an illustration of the issue: http://nbviewer.ipython.org/6488376

If I change it to Gamma(10,1) it works fine...

**Comment by kforeman** [comment]
Did some testing and found that the error occurs for any value of alpha <= 1.0. beta has no effect.

Notebook: http://nbviewer.ipython.org/6490621

**Comment by jsalvatier** [comment]
This is because the mode is 0 for alpha < 1
http://en.wikipedia.org/wiki/Gamma_distribution

It would be nice if the scipy optimization functions would actually find
this value, but they don't seem to. Any suggestions on what kind of
error/behavior would have been more informative?

**Comment by twiecki** [comment]
Seems like it might be related to https://github.com/pymc-devs/pymc/issues/289.

For some reason the optimization is not good at backtracking once it walked out of the support.

**Comment by kforeman** [comment]
Ah, interesting. I think Gamma(1,1) might come up relatively often, since it corresponds to Gelman's Inv-Gamma(1,1) uninformative prior on variances for hierarchical models. 

In my case I'd just substitute alpha=1.0001, but that of course gets dicey as a general rule. 

I'd vote for an error message along the lines of "Optimized value for tau (-0.00048828) is outside of its support", as it wasn't clear that's what was going on. Especially in models with lots of parameters, you have to backtrack to figure out which array entries for max/logp correspond to which parameters.

**Comment by jsalvatier** [comment]
Yeah, 'outside support' is a good phrase to have in that error message.

**Comment by jsalvatier** [comment]
Right now we're using `scipy.fmin_bfgs` as the default optimizer, and as you guys have both seen, it doesn't deal with boundaries very well at all. I'd love to switch to something more robust, even at the expense of some speed (though it has to have decent asymptotic behavior), but I don't know what to switch to. 

I've brought this up to Scipy before, but without much result: https://github.com/scipy/scipy/issues/2169

**Comment by kforeman** [comment]
`scipy.optimize.fmin_tnc` works for alpha<1 and is significantly faster. I think that might be a good solution.

`scipy.optimize.fmin_l_bfgs_b` could also work by adding in the bounds=[] arguments.

**Comment by kforeman** [comment]
Sorry, meant to say works for alpha=1. For alpha<1, you need to put in the bounds. So `bounds=[(1e-10, inf)]` for this model works.

I suppose we could setup for each distribution what the bounds are and then use the bij map to fill it in automatically.

**Comment by jsalvatier** [comment]
I have had some problems on small problems with `fmin_l_bfgs_b` on other small problems in the past (not sure why). If `fmin_tnc` performs decently on the other examples and is more robust, I'm for it. 

Adding bounds information to the model seems like an intriguing idea. It could also lead to better sampling because you can add bounds handling to HMC. It won't solve the problem completely because it is still possible to have areas of 0 likelihood because of the data or because of other interactions in the model, but it could still be helpful. I would definitely be interested in a contribution like that, but I'm probably not going to tackle it myself in the near future.

**Comment by aflaxman** [comment]
Here is a code snippet for someone who has time to work on this more, from when @kforeman was showing it to me yesterday:

```
bij = mc.DictToArrayBijection(mc.ArrayOrdering(m.vars), theta)

lower = dict(tau=.01)
lower = bij.map(lower)

upper = dict(tau=inf)
upper = bij.map(upper)

bounds = zip(lower, upper)

with m: mc.find_MAP(fmin=scipy.optimize.fmin_l_bfgs_b, bounds=bounds)
```

I couldn't figure out how to stick a min and max into the tau object.  It looks like [tau knows its mean and median in the continuous.py source](https://github.com/pymc-devs/pymc/blob/pymc3/pymc/distributions/continuous.py#L348-L350), but I don't know how to access that.  Do you think this is on the right track?

**Comment by jsalvatier** [comment]
You would just do it right next to the mean/median/mode, just add a line that says `min = 0`. 

That does seem like a good approach. Unfortunately random variables do not have an attached Distribution object. They should, and that's something I want to attach, but I haven't worked on it yet.

Then in find_MAP, you would extract all the bounding information you could find.

**Comment by aflaxman** [comment]
Ah, interesting.  Perhaps we can leave this issue open and revisit it when you have attached the Distribution object to the random variable.  Or do you think it would be worth making a feature request about including bounds in MAP more generally?

**Comment by jsalvatier** [comment]
Can you make a bounds MAP request?

**Comment by jsalvatier** [comment]
@kforeman I tried to improve the error message a little bit in 8f8b9bba37e422b1c7d795d6095ce07e80cc2a6f . Can you think of anyway to make it better?

**Comment by twiecki** [comment]
@jsalvatier ideally it'll tell you which variable but I guess we don't have that information currently. Maybe just output the current state of all variables?

**Comment by kforeman** [comment]
@jsalvatier I saw that yesterday, definitely an improvement. I'm thinking the same as @twiecki, showing which variable. Could it be based off of which variables have infinite logp? That way it could be implemented before bounds information is added to distributions.

**Comment by jsalvatier** [comment]
I've tried to give a lot better information here: https://github.com/pymc-devs/pymc/pull/411/files#diff-698344f81aad6681becb2ea365191079R5

**Comment by jsalvatier** [comment]
This should be fixed with #417
